patterns/observers/observer-delegate.cs

```diff
@@ -0,0 +1,82 @@
+// Generic Delegate Implementation of the Observer Pattern
+
+/*
+References used: 
+https://www.youtube.com/watch?v=yoAOo4dM2pk
+https://www.c-sharpcorner.com/article/design-pattern-with-net-delegates/
+https://www.devleader.ca/2023/11/17/examples-of-the-observer-pattern-in-c-how-to-simplify-event-management/#google_vignette
+*/
+
+using System;
+
+// Define the generic delegate for the event handler
+public delegate void EventHandler<T>(T data);
+
+// Define the Observer interface
+public interface IObserver<T>
+{
+    void Update(T data);
+}
+
+// Define the Subject class
+public class Subject<T>
+{
+    // Event 'DataChanged' to which observers can subscribe
+    public event EventHandler<T> DataChanged;
+
+    // Method to notify observers when state of the subject changes
+    public void Notify(T data)
+    {
+        DataChanged?.Invoke(data);
+    }
+}
+
+// Define an example data class
+public class Data
+{
+    public string Message { get; set; }
+}
+
+// Implements the Observer Interface
+public class ConcreteObserver : IObserver<Data>
+{
+    private string name;
+
+    public ConcreteObserver(string name)
+    {
+        this.name = name;
+    }
+
+    public void Update(Data data)
+    {
+        Console.WriteLine($"{name} received message: {data.Message}");
+    }
+}
+
+class Program
+{
+    static void Main(string[] args)
+    {
+        // Create subject
+        Subject<Data> subject = new Subject<Data>();
+
+        // Create observers
+        IObserver<Data> observer1 = new ConcreteObserver("Observer 1");
+        IObserver<Data> observer2 = new ConcreteObserver("Observer 2");
+
+        // Subscribe observers to the subject
+        subject.DataChanged += observer1.Update;
+        subject.DataChanged += observer2.Update;
```

Comment:
I'm not overly familiar with C#, so I had to do a bit of research to see how the `+=` operator works (it's pretty clear from your code what is happening, but I wanted to find a good reference just to make sure I understood). 

C#'s positions as a language that is intended to support .Net means that there are many great features in the language that support event-based programming. This operator is one great example. However, it may be unclear to those less familiar with C# (like me!) or to those who have less programming experience overall. So a bit more explanation (in comments) here would help.

Reply:
We use += or -= operator to subscribe and unsubscribe the event.

Here is the reference link of description:

Reply:
Great, thanks @alinaghale88  - can you make sure to add this to the list of references that you have provided in the header comment in your file?

Reply:
@ash-teach reference link added